### PR TITLE
Feature/cu 89xcek execution phase

### PR DESCRIPTION
**Pull request by kangyee2408**
- Roughly did the basic steps for generating the template
- Some of the functions are modified or taken from Ernest's original cyanprint code
- Currently only downloads NPM modules when it comes to dependencies

### CODE REVIEW

**Comment by AtomiCloudAdmin** [comment]
Task linked: [CU-89xcek Execution Phase](https://app.clickup.com/t/89xcek)

File: src/classLibrary/Parser.ts [hunk 1]
```diff
@@ -11,7 +11,7 @@ import { Utility } from "./Utility/Utility";
 import { Bar, Presets } from "cli-progress";
 
 //taken from kirinnee/CyanPrint (to be edited later if needed)
-class Parser {
+export class Parser {
```
Comment (Lucria) [inline src/classLibrary/Parser.ts]: I think instead of putting export here, can just put `export { Parser }` at the bottom of the file. 😁

File: src/classLibrary/Generator.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import path from "path";
+import { GlobFactory } from "./Utility/GlobFactory";
+import chalk from "chalk";
+import { FileFactory } from "./Utility/FileFactory";
+
+import deleteEmpty from "delete-empty";
+import { PackageResolver } from "./ParsingStrategies/PackageResolver";
+import { PluginHandler } from "./PluginHandler";
+
+export class Generator {
```
Comment (Lucria) [inline src/classLibrary/Generator.ts]: I think we can change the name of this class to just Executor instead to match the name of the phase, but the method names can remain the same

File: src/classLibrary/ParsingStrategies/PackageResolver.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+class PackageResolver implements IParsingStrategy {
+    private readonly util: Utility;
+
+    constructor(util: Utility) {
+        this.util = util;
+    }
+
+    ResolveContents(cyan: CyanSafe, virtualFiles: VirtualFileSystemInstance[]): VirtualFileSystemInstance[] {
+        const flags: CyanFlag = cyan.flags; 
+        if (flags["packages"] != null) {
```
Comment (Lucria) [inline src/classLibrary/ParsingStrategies/PackageResolver.ts]: I think we don't need PackageResolver anymore, all the ParsingStategies should be sufficient already. 

PackageResolver was used in the Old Cyanprint because it was targeted at Javascript projects only, so there was a requirement to be able to parse `package.json` files.

The new Cyanprint aims to be a templating engine beyond Javascript, so we can ignore `PackageResolver` first. It could be useful as a Plugin in the future! 

So basically the entire `PackageResolver` can be removed, but the implementation here I think you can save it somewhere (Maybe in a branch, a stash etc.) as it may be brought back as a Plugin 😄

File: src/classLibrary/Generator.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    
+    async GenerateVFS(cyanSafe: CyanSafe, globFactory: GlobFactory): Promise<VirtualFileSystemInstance[]> {
+    
+        //package resolver may need to think how to expand it to beyond npm modules
+        let strategies: IParsingStrategy[] = [
+            new GuidResolver(this.guidGenerator, this.util),
+            new IfElseResolver(this.util),
+            new InlineFlagResolver(this.util),
+            new VariableResolver(this.util),
+            new PackageResolver(this.util)
```
Comment (Lucria) [inline src/classLibrary/Generator.ts]: Same as below, should remove PackageResolver cause there's no need for it

File: src/classLibrary/Generator.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        //Generate warning of possible possible unaccounted flags
+        parser.CountPossibleRemains(files);
+
+        return files;
+    }
+
+    async GenerateTemplate(cyanSafe: CyanSafe, templatePath: string, folderName: string) {
+        //dest file path
+        //cwd is where the user is calling from
+        let destPath: string = path.resolve(process.cwd(), folderName);
```
Comment (Lucria) [inline src/classLibrary/Generator.ts]: Can be `const`

File: src/classLibrary/Generator.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+        //Check if the target path is empty
+        if (fs.existsSync(destPath)) {
+            //ask if want to delete existing stuff
```
Comment (Lucria) [inline src/classLibrary/Generator.ts]: Can just add a TODO here 😆 

Implementation of delete existing stuff can just follow the oldCyanPrint

File: src/classLibrary/Generator.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            new PackageResolver(this.util)
+        ];
+
+        let parser: Parser = new Parser(this.util, strategies, cyanSafe);
+
+        console.log(chalk.greenBright("Preparation done!"));
+        console.log(chalk.cyanBright("Performing variable and flag scans..."));
+
+        //create globs (empty content vfs)
+        let filesMetadata: IFileSystemInstanceMetadata[] = cyanSafe.globs.Map((g: Glob) => globFactory.GenerateFilesMetadata(g, g.root)).Flatten();
```
Comment (Lucria) [inline src/classLibrary/Generator.ts]: I think instead of returning a `IFileSystemInstanceMetadata` this should return a `VirtualFileSystemInstance` instead, with reference to the method `CreateFileSystemInstance` under `RootFileFactory.ts` in old cyanprint.

File: src/classLibrary/interfaces/interfaces.ts [hunk 8]
```diff
@@ -111,20 +111,31 @@ interface Ignore {
...
+
 // TODO may need further review
 interface IGlobFactory {
-    GenerateFiles(glob: Glob, target: string): IFileSystemInstanceMetadata[];
+    GenerateFiles(glob: GlobSafe, targetDirFromDestRoot?: string): VirtualFileSystemInstance[];
 
     // Callback is used to bump progress
-    ReadFiles(files: IFileSystemInstanceMetadata[], callback?: Function): Promise<FileSystemInstance[]>;
+    ReadFiles(files: VirtualFileSystemInstance[]): Promise<VirtualFileSystemInstance[]>;
 }
```
Comment (Lucria) [inline src/classLibrary/interfaces/interfaces.ts:124]: @kirinnee Do you think we can remove callback here?

File: src/classLibrary/Utility/GlobFactory.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    /**
+	 * Generates the empty file according to the root source path, root destination path, glob and specified target
+	 * returns the files
+	 * @param glob Glob from CyanObject from cyan.script.js
+     * @param targetDirFromDestRoot Optional filepath where the contents of the glob root should be generated at the destination root filepath
+	 */
+    GenerateFiles(glob: GlobSafe, targetDirFromDestRoot: string = './'): VirtualFileSystemInstance[] {
+        let pattern = path.resolve(this.fileFactory.FromRoot, glob.root, glob.pattern.join(''));
```
Comment (Lucria) [inline src/classLibrary/Utility/GlobFactory.ts:33]: @kirinnee Correct me if i'm wrong here

I think `glob.pattern` as an array means that the Glob may have multiple patterns to match different sets of files.  So I don't think a `join('')` will work here.

File: src/classLibrary/interfaces/interfaces.ts [hunk 10]
```diff
@@ -111,20 +111,31 @@ interface Ignore {
     custom: object; // Reserved for custom parsing strategies from plugins
 }
 
+interface GlobSyncOptions {
+    dot?: boolean,
+    ignore?: string | string[]
```
Comment (Lucria) [inline src/classLibrary/interfaces/interfaces.ts:116]:
```
if (glob.ignore != null) {
	opts.ignore = glob.ignore;
}
```

Since `opts.ignore` is equivalent to `glob.ignore`, can just be `string[]`

File: src/classLibrary/Utility/FileFactory.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        let ignoreConfig: Ignore = {
+            variableResolver:  {},
+            inlineResolver: {},
+            ifElseResolver: {},
+            guidResolver: {},
+            custom: {}
+        }
+        if (ignore !== undefined) {
+            ignoreConfig = ignore;
+        }
```
Comment (Lucria) [inline src/classLibrary/Utility/FileFactory.ts]: I think for this portion, we can use lodash [`defaultsDeep` ](https://lodash.com/docs/4.17.15#defaultsDeep)

File: src/classLibrary/Executor.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+                File: (file: FileSystemInstance) => {
+                    return (file.metadata.sourceAbsolutePath.ReplaceAll("\\\\", "/").split("/").Last()! !== "yarn.lock");
+
+                },
+                Folder: (folder: DirectorySystemInstance) => {
+                    return (folder.metadata.sourceAbsolutePath.ReplaceAll("\\\\", "/").split("/").Last()! !== "yarn.lock");
+                },
+                default: () => false
+            });
+        });
```
Comment (Lucria) [inline src/classLibrary/Executor.ts]: @kirinnee Correct me if i'm wrong also, but I think this portion is not needed anymore (for now at least, these part may return in a plugin)

Cause these portion is for cleaning up the lock files and `node_modules` but in the new Cyanprint, by default we don't consider this and I think should be handled by a plugin?

File: src/classLibrary/Utility/GlobFactory.ts [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    async AWriteFile(files: VirtualFileSystemInstance[]) {
+        let bar: Bar = new Bar({}, Presets.shades_grey);
+        let counter: number = 0;
+        bar.start(files.length, 0);
+        let promises: Promise<void>[] = [];
+        files.Each((f: VirtualFileSystemInstance) => {
+            return VirtualFileSystemInstance.match(f, {
+                File: async (file: FileSystemInstance) => {
+                    let content = file.content;
+                    const isBinary = await isBinaryFile(file.metadata.sourceAbsolutePath);
```
Comment (Lucria) [inline src/classLibrary/Utility/GlobFactory.ts:72]: I think we can use `file.content` to check whether it's a binary or not. @kirinnee Correct me if i'm wrong here:

```
const FileContent = Union({
    String: of<string>(),
    Buffer: of<Buffer>(),
});
```

So if `FileContent` is of type `Buffer` then it is a binary file

File: src/classLibrary/Utility/GlobFactory.ts [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+			readCounter++;
+			readBar.update(readCounter);
+			if (readCounter >= readBar.getTotal()) {
+				readBar.stop();
+			}
+		})));
+		return await Promise.all(promises);
+    }
+
+    async AWriteFile(files: VirtualFileSystemInstance[]) {
```
Comment (Lucria) [inline src/classLibrary/Utility/GlobFactory.ts:63]: I think we can consider moving this method to it's own class/interface called `FileWriter`

File: src/classLibrary/Executor.ts [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        let globFactory: GlobFactory = new GlobFactory(fileFactory, this.util);
+
+        let virtualFSInstances =  await this.GenerateVFS(cyanSafe, globFactory);
+        
+        //Asynchronous write to target directory
+        await globFactory.AWriteFile(virtualFSInstances);
+
+        //use plugin handler
+        let pluginHandler: PluginHandler = new PluginHandler(fileFactory);
+        pluginHandler.DownloadPlugins(cyanSafe, folderName);
```
Comment (Lucria) [inline src/classLibrary/Executor.ts]: I think we missing an `await` here

File: src/classLibrary/PluginHandler.ts [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        console.log(paths);
+        let cds: string[] = paths.map(p => {
+            let destPath = path.resolve(process.cwd(), folderName);
+            let dirOfPackageJson = path.resolve(p, "..");
+            return path.relative(destPath, dirOfPackageJson);
+        });
+        console.log(cds);
+        console.log(chalk.cyanBright("Installing NPM modules"));
+        let reply = "";
+        let hasYarn = await this.ExecuteCommandSimple("yarn", ["-v"], "", true);
```
Comment (Lucria) [inline src/classLibrary/PluginHandler.ts:65]: Oh yeahh we should consider changing our custom `ExecuteCommand` method to using the[ `execa` package.](https://www.npmjs.com/package/execa)

File: src/classLibrary/PluginHandler.ts [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+                    await this.ExecuteCommandSimple("git", ["remote", "add", "origin", git.remote], folderName);
+                    let reply: string = await this.ExecuteCommand("git", ["push", "origin", "--all"], "Added and pushed to remote repository", folderName);
+                    console.log(reply);
+                }
+            }
+        }*/
+    }
+
+    async DownloadNpm(cyanSafe: CyanSafe, folderName: string) {
+        let paths: string[] = cyanSafe.globs.map(glob => this.fileFactory.GetAbsoluteFilePathsOfFileInDestinationPath("package.json", glob.root, glob.pattern as string, glob.ignore)).flat();
```
Comment (Lucria) [inline src/classLibrary/PluginHandler.ts:51]: There are some errors here about type casting `glob.pattern` as `string` cause `glob.pattern` is a `string[]`, is it okay if we use `string[]`?

**Comment by Lucria** [comment]
I just realised that there is this extra chunk here: 

```
(typeof data === 'object' && data !== null && typeof data.valueOf() === 'boolean')
``` 

In `Utility.ts` under `FlattenBooleanValueObject`, what is the purpose of this extra check ah? (I'm not sure so I just want to clarify

File: src/classLibrary/Utility/FileFactory.ts [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+    GetAbsoluteFilePathsOfFileInDestinationPath(fileName: string, fromRoot: string = "./", pattern: string = "**/*.*", ignore?: string | string[]): string[] {
+        let pathPattern = path.resolve(this.ToRoot, fromRoot, pattern);
+		let opts: GlobSyncOptions = {dot: true}
+        if (ignore != null) {
+			opts.ignore = ignore;
+		}
+        //returns all absolute path according to the glob pattern 
+        let files: string[] = _glob.sync(pathPattern, opts);
+        return files.filter((path: string) => path.includes(fileName));
+    }
```
Comment (Lucria) [inline src/classLibrary/Utility/FileFactory.ts:76]: What's the purpose of this method? I don't think we need it under IFileFactory.

For npm plug-in we can think about a better way to implement it later, but I don't think this method should be needed.
